### PR TITLE
Extend string-based logging to C and Fortran clients

### PR DESCRIPTION
**Pull request by billschereriii**

### CODE REVIEW

**Review by ashao — CHANGES_REQUESTED** [review]
Looks good to me. Some minor comments to address, but nothing particularly onerous.

File: doc/changelog.rst [hunk 1]
```diff
@@ -12,6 +12,8 @@ This section details changes made in the development branch that have not yet be
 
 Description
 
+- Add support for string-based logging for C and Fortran clients
```
[the remark this answers is not in the document]
Comment (billschereriii) [reply doc/changelog.rst:15]: done

File: include/c_logger.h [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+*   \param data_len Length in characters of data to be logged
+*/
+void log_error_noexcept(
+    const void* context,
+    SRLoggingLevel level,
+    const char* data,
+    size_t data_len);
+
+/*!
+*   \brief Conditionally log data if the logging level is high enough
```
[the remark this answers is not in the document]
Comment (billschereriii) [reply include/c_logger.h:94]: description expanded

File: src/c/c_logger.cpp [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+// (exception-free variant)
+extern "C" void log_data_noexcept(
+    const void* context,
+    SRLoggingLevel level,
+    const char* data,
+    size_t data_len)
+{
+    try {
+        SR_CHECK_PARAMS(context != NULL && data != NULL);
+        const SRObject* ctxt = reinterpret_cast<const SRObject*>(context);
```
[the remark this answers is not in the document]
Comment (billschereriii) [reply src/c/c_logger.cpp:56]: renamed

File: src/fortran/logger.F90 [hunk 4]
```diff
@@ -42,15 +42,18 @@ module smartredis_logger
 #include "logger/logger_interfaces.inc"
 
 interface log_data
-  module procedure log_data_client, log_data_dataset, log_data_logcontext
+  module procedure log_data_client, log_data_dataset, &
+    log_data_logcontext, log_data_string
```
[the remark this answers is not in the document]
Comment (billschereriii) [reply src/fortran/logger.F90:46]: aligned

File: src/fortran/logger.F90 [hunk 5]
```diff
@@ -119,14 +122,38 @@ subroutine log_data_logcontext(context, level, data)
...
+subroutine log_data_string(context, level, data)
+  character(len=*),         intent(in) :: context !< Context for logging
+  integer (kind=enum_kind), intent(in) :: level   !< Minimum logging level to log the data at
+  character(len=*),         intent(in) :: data    !< Data to be logged
+
+  ! Local variables
+  character(kind=c_char, len=:), allocatable :: c_context, c_data
+  integer(kind=c_size_t) :: c_context_length, c_data_length
+
+  allocate(character(kind=c_char, len=len_trim(context)) :: c_context)
```
[the remark this answers is not in the document]
Comment (billschereriii) [reply src/fortran/logger.F90:135]: Removed the allocate calls from logger.F90.  Still a ton of equivalent calls in client.F90 and dataset.F90 -- should probably write up a ticket to get them cleaned up.

**Review by ashao — APPROVED** [review]
Looks great. Thanks Bill!

**Pull request merged**